### ISSUE TITLE
CODING PART

### ISSUE THREAD

**Issue by Raw2003** [Raw2003/Practical#1]
`# -----------------------------------------------------------
# Indian House Price Prediction
# Using Lasso Regression (feature selection) + SVM (SVR)
# With Data Visualization
# Dataset: Indian House Prices.csv
# -----------------------------------------------------------

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import Lasso
from sklearn.svm import SVR
from sklearn.metrics import mean_squared_error, r2_score

# -----------------------------
# 1. Load Dataset
# -----------------------------
dataset_path ="C:\Users\HP\Desktop\SVM_MACHINE PROJECT\Indian House Prices.csv"  
df = pd.read_csv(dataset_path)

print("===== Dataset Info =====")
print("Shape:", df.shape)
print("\nColumns:", df.columns.tolist())
print("\nFirst 5 rows:")
print(df.head())

if "Unnamed: 0" in df.columns:
    df.drop(columns=["Unnamed: 0"], inplace=True)

print("\nShape after dropping Unnamed: 0 (if existed):", df.shape)

# -----------------------------
# 2. Handle Missing Values
# -----------------------------
print("\n===== Missing Values (before filling) =====")
print(df.isnull().sum())

# Separate numeric and categorical columns
numeric_cols = df.select_dtypes(include=["int64", "float64"]).columns
categorical_cols = df.select_dtypes(include=["object"]).columns

# Fill numeric NaNs with median
df[numeric_cols] = df[numeric_cols].fillna(df[numeric_cols].median())

# Fill categorical NaNs with mode (most frequent)
if len(categorical_cols) > 0:
    df[categorical_cols] = df[categorical_cols].fillna(df[categorical_cols].mode().iloc[0])

print("\n===== Missing Values (after filling) =====")
print(df.isnull().sum())

# -----------------------------
# 3. Encode Categorical Features
# -----------------------------
print("\nCategorical Columns:", categorical_cols.tolist())

df_encoded = pd.get_dummies(df, columns=categorical_cols, drop_first=True)

print("\nShape after One-Hot Encoding:", df_encoded.shape)

# -----------------------------
# 4. Split Features & Target
# -----------------------------
target_col = "Price"   # target variable

X = df_encoded.drop(target_col, axis=1)
y = df_encoded[target_col]

print("\nTotal Features before Lasso:", X.shape[1])

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

# -----------------------------
# 5. Scale Features (for Lasso & SVM)
# -----------------------------
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# -----------------------------
# 6. Lasso Regression (Feature Selection)
# -----------------------------
lasso_alpha = 0.01  # you can tune this value
lasso = Lasso(alpha=lasso_alpha, random_state=42, max_iter=10000)
lasso.fit(X_train_scaled, y_train)

coefficients = lasso.coef_
feature_names = X.columns

feature_importance = pd.DataFrame({
    "Feature": feature_names,
    "Coefficient": coefficients
})

# Keep only non-zero coefficient features
important_features = feature_importance[feature_importance["Coefficient"] != 0]
important_features_sorted = important_features.reindex(
    important_features["Coefficient"].abs().sort_values(ascending=False).index
)

print("\n===== Important Features Selected by Lasso (non-zero coefficients) =====")
print(important_features_sorted)

print("\nTotal Selected Features by Lasso:", important_features_sorted.shape[0])

# Boolean mask of selected features
important_mask = coefficients != 0

# Reduce scaled feature sets for SVM using only Lasso-selected features
X_train_lasso = X_train_scaled[:, important_mask]
X_test_lasso = X_test_scaled[:, important_mask]

selected_feature_names = feature_names[important_mask]
print("\nSelected Feature Names for SVM:")
print(selected_feature_names.tolist())

# -----------------------------
# 7. SVM Regression (SVR) on Selected Features
# -----------------------------
svm_model = SVR(kernel="rbf")
svm_model.fit(X_train_lasso, y_train)

y_pred = svm_model.predict(X_test_lasso)

rmse = np.sqrt(mean_squared_error(y_test, y_pred))
r2 = r2_score(y_test, y_pred)

print("\n===== SVM Model Performance (on Lasso-selected features) =====")
print(f"RMSE: {rmse:.4f}")
print(f"R2 Score: {r2:.4f}")

# -----------------------------
# 8. DATA VISUALIZATION
# -----------------------------
sns.set_style("whitegrid")

# ---------- Plot 1: Distribution of House Prices ----------
plt.figure(figsize=(8, 5))
sns.histplot(y, kde=True)
plt.title("Distribution of House Prices")
plt.xlabel("Price")
plt.ylabel("Frequency")
plt.tight_layout()
plt.show()

# ---------- Plot 2: Price vs Area ----------
if "Area" in df.columns:
    plt.figure(figsize=(8, 5))
    plt.scatter(df["Area"], df["Price"], alpha=0.4)
    plt.title("Price vs Area")
    plt.xlabel("Area (sq.ft)")
    plt.ylabel("Price")
    plt.tight_layout()
    plt.show()

# ---------- Plot 3: Price vs Number of Bedrooms (Box Plot) ----------
if "No. of Bedrooms" in df.columns:
    plt.figure(figsize=(8, 5))
    # Convert to int for cleaner x-axis if needed
    bedrooms = df["No. of Bedrooms"].astype(int)
    sns.boxplot(x=bedrooms, y=df["Price"])
    plt.title("Price vs Number of Bedrooms")
    plt.xlabel("No. of Bedrooms")
    plt.ylabel("Price")
    plt.tight_layout()
    plt.show()

# ---------- Plot 4: Lasso Feature Importance (Top 10) ----------
plt.figure(figsize=(10, 5))
top_n = 10
top_features = important_features_sorted.head(top_n)

plt.bar(top_features["Feature"], top_features["Coefficient"])
plt.title(f"Top {top_n} Important Features (Lasso Coefficients)")
plt.xticks(rotation=75, ha="right")
plt.ylabel("Coefficient")
plt.tight_layout()
plt.show()

# ---------- Plot 5: Actual vs Predicted Prices ----------
plt.figure(figsize=(7, 7))
plt.scatter(y_test, y_pred, alpha=0.5)
# Perfect prediction line
min_price = min(y_test.min(), y_pred.min())
max_price = max(y_test.max(), y_pred.max())
plt.plot([min_price, max_price], [min_price, max_price], linestyle="--")
plt.title("Actual vs Predicted House Prices (SVM)")
plt.xlabel("Actual Price")
plt.ylabel("Predicted Price")
plt.tight_layout()
plt.show()

print("\nProject finished successfully!")
`
